File: Leetcode Challenge/09_September_2020/Python/Week 4/4_Largest Number.py
#Given a list of non negative integers, arrange them such that they form the largest number.
#
#Example 1:
#
#Input: [10,2]
#Output: "210"
#Example 2:
#
#Input: [3,30,34,5,9]
#Output: "9534330"
#Note: The result may be very large, so you need to return a string instead of an integer.

class CustomCompKey(str):
    def __lt__(x, y):
        return x + y > y + x

class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        nums = map(str, nums)
        cmp = lambda x, y : -1 if x + y > y + x else 1 if x + y < y + x else 0
        nums = sorted(nums, key = cmp_to_key(cmp))
        #nums = sorted(nums, key = CustomCompKey)
        ans = ''.join(nums)
        return '0' if ans[0] == '0' else ans

File: Leetcode Challenge/09_September_2020/C++/Week 4/4_Largest Number.cpp
//Given a list of non negative integers, arrange them such that they form the largest number.
//
//Example 1:
//
//Input: [10,2]
//Output: "210"
//Example 2:
//
//Input: [3,30,34,5,9]
//Output: "9534330"
//Note: The result may be very large, so you need to return a string instead of an integer.

class Solution {
public:
    
    static bool cmp(string &x, string &y) {
        return x + y > y + x;
    }
    
    string largestNumber(vector<int>& nums) {
        vector<string> s(nums.size(), "");
        string ans = "";
        for(int i = 0; i < nums.size(); i++) {
            s[i] = to_string(nums[i]);
        }
        sort(s.begin(), s.end(), cmp);
        if(s[0] == "0") {
            return "0";
        } else {
            for(string e : s) {
                ans += e;
            }
            return ans;
        }
    }
};